src/fast_convolver.rs

```diff
@@ -8,24 +20,259 @@ pub enum ConvolutionMode {
     FrequencyDomain { block_size: usize },
 }
 
+fn zero_pad<T: Default + Clone + Copy>(input: &[T], num_zeros: usize) -> Vec<T> {
+    let mut v = vec![T::default(); input.len() + num_zeros];
+    for i in 0..input.len() {
+        v[i] = input[i];
+    }
+    return v;
+}
+
+fn build_filter_bank(impulse_response: Vec<f32>, block_size: usize) -> Vec<Vec<Complex32>> {
+    let mut planner = FftPlanner::new();
+    let fft = planner.plan_fft_forward(block_size * 2);
+    let impulse_response: Vec<Complex32> = impulse_response.into_iter().map(|e| Complex32::from(e)).collect();
+
+    let mut output = Vec::new();
+    for i in 0..impulse_response.len()/block_size {
+        let mut filter: Vec<Complex32> = vec![Complex32::new(0.0, 0.0); block_size * 2];
+        filter[..block_size].clone_from_slice(&impulse_response[i*block_size..(i+1)*block_size]);
+        fft.process(&mut filter);
+        output.push(filter[..block_size+1].to_vec());
+    }
+    return output;
+}
+
 impl FastConvolver {
     pub fn new(impulse_response: &[f32], mode: ConvolutionMode) -> Self {
-        todo!("implement")
+        let mut block_size = 1;
+        let mut n_filters = 1;
+
+        let vec_impulse: Vec<f32> = match mode {
+            ConvolutionMode::FrequencyDomain { block_size: b_s } => {
+                block_size = b_s;
+                //Add some zeros to make impulse response a nice multiple of block size
+                let nzeros = block_size - impulse_response.len() % block_size;
+                n_filters = (impulse_response.len() + nzeros) / block_size;
+                zero_pad(impulse_response, nzeros)
+            },
+            _ => impulse_response.to_vec(),
+        };
+
+        let impulse_len = vec_impulse.len();
+        FastConvolver {
+            mode,
+
+            impulse_response: vec_impulse.clone(),
+            previous_block: vec![0f32; impulse_len],
+
+            freq_domain_delay_line: vec![Complex32::new(0.0, 0.0); (block_size + 1) * n_filters],
+            time_domain_input_buffer: vec![0.0; block_size * 2],
+            filter_bank: build_filter_bank(vec_impulse, block_size),
+            block_size
+        }
     }
 
     pub fn reset(&mut self) {
         todo!("implement")
     }
```

Comment:
I also had difficulty with finding variables that needed to be reset. If you use RingBuffer for your convolution processes then you could reset your use of the ring_buffer